### ISSUE TITLE
segfault with buffer larger than INT_MAX

### ISSUE THREAD

**Issue by yfguo** [pmodels/mpich#3546]
The bug is reported on the MPICH-discuss mailing list as follow.

```
Hi all,

we found another integer overflow when using MPI_Sendrecv_replace, see
attached file.
When the buffer size exceeds MAX_INT, a segmentation fault occurs.
I could reproduce this with the MPICH 3.3a2 installation on my Ubuntu
desktop.

@Jeff: I could reproduce the segfault as well as with IntelMPI 2018
IntelMPI 2019 detects a message truncation


Best
Joachim
--
Dipl.-Inf. Joachim Protze

IT Center
Group: High Performance Computing
Division: Computational Science and Engineering
RWTH Aachen University
Seffenter Weg 23
D 52074  Aachen (Germany)
Tel: +49 241 80- 24765
Fax: +49 241 80-624765
protze@itc.rwth-aachen.de
www.itc.rwth-aachen.
```

Code for reproducing the bug. `sendrecv_replace.c`
```c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
int main(void)
{
  int rank, size,i;
  MPI_Init(NULL,NULL);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);
  double* buf = malloc(323262400L*sizeof(double));

  for(i=(1<<28)-4; i< (1<<28)+4; i++){
    printf("Size: %i\n", i);
    MPI_Sendrecv_replace(buf, i, MPI_DOUBLE, size-rank-1, 1111, size-rank-1, 1111, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
  }
  MPI_Sendrecv_replace(buf, 323262400, MPI_DOUBLE, size-rank-1, 1111, size-rank-1, 1111, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

  MPI_Finalize();
  return 0;
}
```

**Comment by yfguo** [comment]
I have tested the latest `master` and confirmed the reported behavior.

**Comment by sudheer9** [comment]
@hzhou I could not reproduce the reported behavior with the latest `master`.

**Comment by hzhou** [comment]
@sudheer9 Thanks. Could you try a few releases, e.g. v3.3.2, v3.4, v4.0a1, to get an idea at which point we fixed the bug?

**Comment by sudheer9** [comment]
@hzhou tested this test program with v3.3.2, v3.4.1 and v4.0a1. The code segfault with v3.3.2 and works fine with v3.4.1 and later.

**Comment by hzhou** [comment]
> 
> 
> @hzhou tested this test program with v3.3.2, v3.4.1 and v4.0a1. The code segfault with v3.3.2 and works fine with v3.4.1 and later.

Thanks! With that message, you can close this issue.

**Comment by sudheer9** [comment]
The reported `segfault` behavior tested to be fixed from v3.4.1 and later MPICH releases.
